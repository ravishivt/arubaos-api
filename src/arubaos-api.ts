import { get, OptionsWithUrl, post } from 'request-promise-native';

export class ArubaOsApiClient {
  public sessionCookie?: string;
  private options: Required<ArubaOsApiOptions>;

  constructor(options: ArubaOsApiOptions) {
    this.options = { ...this.getDefaultOptions(), ...options };
  }

  public async login() {
    try {
      const response = await post({
        url: `${this.getBaseUrl()}/api/login`,
        headers: { 'Content-Type': 'application/json;' },
        body: `username=${this.options.username}&password=${this.options.password}`,
        strictSSL: this.options.strictSSL,
        resolveWithFullResponse: true,
      });
      const responseJSON = JSON.parse(response.body);
      this.sessionCookie = responseJSON._global_result.UIDARUBA;
      return responseJSON;
    } catch (err) {
      if (err.message) {
        err = err.message;
      }
      throw new Error('Failed to login. ' + err);
    }
  }

  public async logout() {
    try {
      const response = await get({
        url: `${this.getBaseUrl()}/api/logout`,
        headers: { 'Content-Type': 'application/json;' },
        strictSSL: this.options.strictSSL,
        resolveWithFullResponse: true,
      });
      this.sessionCookie = undefined;
      return JSON.parse(response.body);
    } catch (err) {
      if (err.message) {
        err = err.message;
      }
      throw new Error('Failed to log out ' + err);
    }
  }

  /**
   * Generic helper method to prepare API requests with all supported
   *   options including GET and SET requests.
   */
  public async apiRequest({
    requestType,
    configPath,
    objectName,
    payload,
    getModifiers,
  }: ArubaOsApiRequestOptions) {
    if (!this.sessionCookie) {
      throw new Error(
        'ERROR: Authentication cookie not set.  Please use the login() method before making an API request.',
      );
    }
    try {
      if (payload && getModifiers) {
        throw new Error(
          'Cannot specify GET modifiers for POST requests (requests with a payload).',
        );
      }
      const getModifierParams = this.processGetModifiers(getModifiers);
      const preparedUrl = `${this.getBaseUrl()}/configuration/${requestType}${
        objectName ? `/${objectName}` : ''
      }?${getModifierParams}UIDARUBA=${this.sessionCookie}&config_path=${encodeURIComponent(
        configPath || this.options.defaultConfigPath,
      )}`;
      if (this.options.debugEnabled) {
        console.debug('ArubaOS-API [DEBUG]: preparedUrl', preparedUrl);
      }
      const requestOptions: OptionsWithUrl = {
        url: preparedUrl,
        ...(payload ? { body: JSON.stringify(payload) } : {}),
        headers: {
          'Content-Type': 'application/json;',
          Cookie: `SESSION=${this.sessionCookie}`,
        },
        strictSSL: this.options.strictSSL,
        resolveWithFullResponse: true,
      };
      /**
       * If `payload` param is passed, use a `POST` requst and put the payload in the request body.
       * Else, use `GET` with no payload.
       */
      const response = await (payload ? post(requestOptions) : get(requestOptions));
      if (response.headers && response.headers['content-type'] !== 'application/json') {
        throw new Error(
          `ERROR: Content-Type received is not 'application/json', it is '${response.headers['content-type']}'.  ` +
            response.body,
        );
      }
      return JSON.parse(response.body);
    } catch (err) {
      if (err.message) {
        err = err.message;
      }
      throw new Error(
        `API request failed for "/${requestType}${objectName ? `/${objectName}` : ''}". ${err}`,
      );
    }
  }

  private processGetModifiers(getModifiers?: ArubaOsApiRequestGetModifiers) {
    if (!getModifiers) {
      return '';
    }

    let [filterParam, sortParam, countParam, dataTypeParam, paginateParams] = [
      '',
      '',
      '',
      '',
      '',
      '',
    ];
    if (getModifiers.filter) {
      filterParam = `filter=${JSON.stringify(getModifiers.filter)}&`;
    }
    if (getModifiers.sort) {
      sortParam = `sort=${getModifiers.sort.oper}${getModifiers.sort.key}&`;
    }
    if (getModifiers.count) {
      countParam = `count=${getModifiers && getModifiers.count}&`;
    }
    if (getModifiers.dataType) {
      dataTypeParam = `type=${getModifiers && getModifiers.dataType.join(',')}&`;
    }
    if (getModifiers.paginate) {
      paginateParams =
        `offset=${getModifiers.paginate.offset}&` +
        `limit=${getModifiers.paginate.limit}&` +
        (getModifiers.paginate.total ? `total=${getModifiers.paginate.total}&` : '');
    }
    return filterParam + sortParam + dataTypeParam + countParam + paginateParams;
  }

  private getBaseUrl() {
    return `https://${this.options.host}:${this.options.port}/v1`;
  }

  private getDefaultOptions(): Required<ArubaOsApiOptions> {
    return {
      port: 4343,
      strictSSL: true,
      defaultConfigPath: '/mm/mynode',
      debugEnabled: false,
    } as Required<ArubaOsApiOptions>;
  }
}

export type FilterOperKeys = '$eq' | '$neq' | '$gt' | '$gte' | '$lt' | '$lte' | '$in' | '$nin';

export interface ArubaOsApiOptions {
  // Management IP of the controller the API will target.
  host: string;
  // Admin user on the controller.
  username: string;
  // Consider using environment variable or command line argument for added security.
  password: string;
  // Set if using non-default port (4343).
  port?: number;
  // The default target config-node or config-path.  This can be overriden later in code.
  defaultConfigPath?: string;
  // Disable if controller doesn't have a trusted certificate.
  strictSSL?: boolean;
  // Enable to get additional logging.
  debugEnabled?: boolean;
}

export interface ArubaOsApiRequestGetModifiers {
  /**
   * Limit which objects, sub_objects, or configuration elements should be
   *   present in the response.
   */
  filter?: Array<{
    [key: string]: {
      [key in FilterOperKeys]?: Array<string | number | boolean> | string | number | boolean;
    };
  }>;
  /**
   * Data from the GET request can be sorted based on a single field
   */
  sort?: { oper: '+' | '-'; key: string };
  /**
   * Returns the total count of an object for multi-instance object or for
   *   multi-instance subobject rather than the actual details of the objects.
   */
  count?: number;
  /**
   * What type of data should be returned.
   */
  dataType?: Array<
    | 'non-default'
    | 'default'
    | 'local'
    | 'user'
    | 'system'
    | 'pending'
    | 'committed'
    | 'inherited'
    | 'meta-n-data'
    | 'meta-only'
  >;
  /**
   * Get results in chunks at a time.
   * See docs for more details on the `total` (aka `count`) param.
   */
  paginate?: {
    limit: number;
    offset: number;
    total?: number;
  };
}

export interface ArubaOsApiRequestOptions {
  requestType: 'object' | 'container';
  /**
   * Name of a specific API `object` or `container`.
   * Don't specify a value if you want a list of all objects or containers.
   */
  objectName?: string;
  /**
   * Data to be sent in the Set (POST) request.
   * If a value is not passed, the request will be a GET request.
   */
  payload?: { [key: string]: any };
  /**
   * Override the DEFAULT_CONFIG_PATH setting for this request.
   */
  configPath?: string;
  getModifiers?: ArubaOsApiRequestGetModifiers;
}
