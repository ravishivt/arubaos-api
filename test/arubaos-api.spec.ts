import * as request from 'request-promise-native';

import { ArubaOsApiClient, ArubaOsApiOptions } from '../src/arubaos-api';

jest.mock('request-promise-native');
const mockedRequest: jest.Mocked<typeof request> = request as any;

const testSetUp = (options?: Partial<ArubaOsApiOptions>) => {
  return new ArubaOsApiClient({
    host: '1.2.3.4',
    username: 'admin',
    password: 'abc',
    debugEnabled: false,
    ...options,
  });
};

describe(`ArubaOsApiClient`, () => {
  describe(`login`, () => {
    it(`should error if the request throws an error`, async () => {
      const aosClient = testSetUp();

      mockedRequest.post.mockImplementation(() => {
        throw new Error('timeout');
      });
      try {
        await aosClient.login();
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.post).toHaveBeenCalledTimes(1);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should successfully login with correct credentials`, async () => {
      const aosClient = testSetUp();

      mockedRequest.post.mockResolvedValue({
        body: JSON.stringify({
          _global_result: {
            UIDARUBA: 'AOS-COOKIE',
            status: '0',
            status_str: "You've logged in successfully.",
          },
        }),
      });
      const response = await aosClient.login();
      expect(response).toMatchSnapshot();
      expect(aosClient.sessionCookie).toBe('AOS-COOKIE');
      expect(mockedRequest.post).toHaveBeenCalledTimes(1);
    });
  });

  describe(`logout`, () => {
    it(`should error if the request throws an error`, async () => {
      const aosClient = testSetUp();

      mockedRequest.get.mockImplementation(() => {
        throw new Error('timeout');
      });
      try {
        await aosClient.logout();
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.get).toHaveBeenCalledTimes(1);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should successfully logout`, async () => {
      const aosClient = testSetUp();

      mockedRequest.get.mockResolvedValue({
        body: JSON.stringify({
          _global_result: {
            UIDARUBA: '(null)',
            status: '0',
            status_str: "You've been logged out successfully.",
          },
        }),
      });
      const response = await aosClient.logout();
      expect(response).toMatchSnapshot();
      expect(aosClient.sessionCookie).toBe(undefined);
      expect(mockedRequest.get).toHaveBeenCalledTimes(1);
    });
  });

  describe(`apiRequest`, () => {
    it(`should error if sessionCookie is not set`, async () => {
      const aosClient = testSetUp();

      try {
        await aosClient.apiRequest({
          requestType: 'object',
        });
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.get).toHaveBeenCalledTimes(0);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should error if the request throws an error`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockImplementation(() => {
        throw new Error('timeout');
      });
      try {
        await aosClient.apiRequest({
          requestType: 'object',
        });
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.get).toHaveBeenCalledTimes(1);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should error if "payload" and "getModifiers" are passed`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      try {
        await aosClient.apiRequest({
          requestType: 'object',
          getModifiers: {
            count: 5,
          },
          payload: { testProp: 'newValue' },
        });
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.get).toHaveBeenCalledTimes(0);
        expect(mockedRequest.post).toHaveBeenCalledTimes(0);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should error if response "Content-Type" isn't "application/json"`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockResolvedValue({ headers: { 'content-type': 'text/html' }, body: '{}' });
      try {
        await aosClient.apiRequest({
          requestType: 'object',
        });
        fail('Should have thrown an exception');
      } catch (err) {
        expect(mockedRequest.get).toHaveBeenCalledTimes(1);
        expect(err).toMatchSnapshot();
      }
    });

    it(`should successfully make a get object request`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockResolvedValue({ body: '{}' });
      const response = await aosClient.apiRequest({
        requestType: 'object',
        objectName: 'testObject',
      });
      expect(response).toEqual({});
      expect(mockedRequest.get).toHaveBeenCalledTimes(1);
      expect(mockedRequest.post).toHaveBeenCalledTimes(0);
      expect(mockedRequest.get.mock.calls[0][0]).toMatchSnapshot();
    });

    it(
      `should successfully make a get object request and print out ` +
        `the "preparedUrl" when "debugEnabled: true"`,
      async () => {
        const aosClient = testSetUp({ debugEnabled: true });
        aosClient.sessionCookie = 'AOS-COOKIE';

        mockedRequest.get.mockResolvedValue({ body: '{}' });
        jest.spyOn(global.console, 'debug');
        const response = await aosClient.apiRequest({
          requestType: 'object',
        });
        expect(response).toEqual({});
        expect(console.debug).toHaveBeenCalledTimes(1);
        expect(mockedRequest.get).toHaveBeenCalledTimes(1);
        expect(mockedRequest.post).toHaveBeenCalledTimes(0);
        expect(mockedRequest.get.mock.calls[0][0]).toMatchSnapshot();
      },
    );

    it(`should successfully make a get object request with an empty "getModifiers"`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockResolvedValue({ body: '{}' });
      const response = await aosClient.apiRequest({
        requestType: 'object',
        getModifiers: {},
      });
      expect(response).toEqual({});
      expect(mockedRequest.get).toHaveBeenCalledTimes(1);
      expect(mockedRequest.post).toHaveBeenCalledTimes(0);
      expect(mockedRequest.get.mock.calls[0][0]).toMatchSnapshot();
    });

    it(`should successfully make a get object request with "count" only in "getModifiers"`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockResolvedValue({ body: '{}' });
      const response = await aosClient.apiRequest({
        requestType: 'object',
        objectName: 'testObject',
        getModifiers: {
          count: 5,
        },
      });
      expect(response).toEqual({});
      expect(mockedRequest.get).toHaveBeenCalledTimes(1);
      expect(mockedRequest.post).toHaveBeenCalledTimes(0);
      expect(mockedRequest.get.mock.calls[0][0]).toMatchSnapshot();
    });

    it(`should successfully make a get object request with "getModifiers"`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.get.mockResolvedValue({ body: '{}' });
      const response = await aosClient.apiRequest({
        requestType: 'object',
        objectName: 'testObject',
        getModifiers: {
          filter: [
            { 'acl_sess.accname': { $nin: ['-acl'] } },
            { OBJECT: { $eq: ['acl_sess.accname'] } },
          ],
          sort: { oper: '+', key: 'testKey' },
          count: 5,
          dataType: ['committed', 'pending'],
          paginate: { limit: 5, offset: 5, total: 5 },
        },
      });
      expect(response).toEqual({});
      expect(mockedRequest.get).toHaveBeenCalledTimes(1);
      expect(mockedRequest.post).toHaveBeenCalledTimes(0);
      expect(mockedRequest.get.mock.calls[0][0]).toMatchSnapshot();
    });

    it(`should successfully make a set object request if "payload" data is passed`, async () => {
      const aosClient = testSetUp();
      aosClient.sessionCookie = 'AOS-COOKIE';

      mockedRequest.post.mockResolvedValue({ body: '{}' });
      const response = await aosClient.apiRequest({
        requestType: 'object',
        objectName: 'testObject',
        payload: { testProp: 'newValue' },
      });
      expect(response).toEqual({});
      expect(mockedRequest.get).toHaveBeenCalledTimes(0);
      expect(mockedRequest.post).toHaveBeenCalledTimes(1);
      expect(mockedRequest.post.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
