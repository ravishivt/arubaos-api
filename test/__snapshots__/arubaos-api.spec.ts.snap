// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArubaOsApiClient apiRequest should error if "payload" and "requestModifiers" are passed 1`] = `[Error: API request failed for "/object". Cannot specify GET modifiers for POST requests (requests with a payload).]`;

exports[`ArubaOsApiClient apiRequest should error if response "Content-Type" isn't "application/json" 1`] = `[Error: API request failed for "/object". ERROR: Content-Type received is not 'application/json', it is 'text/html'.  {}]`;

exports[`ArubaOsApiClient apiRequest should error if sessionCookie is not set 1`] = `[Error: ERROR: Authentication cookie not set.  Please use the login() method before making an API request.]`;

exports[`ArubaOsApiClient apiRequest should error if the request throws an error 1`] = `[Error: API request failed for "/object". timeout]`;

exports[`ArubaOsApiClient apiRequest should successfully make a get object request 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json;",
    "Cookie": "SESSION=AOS-COOKIE",
  },
  "resolveWithFullResponse": true,
  "strictSSL": true,
  "url": "https://1.2.3.4:4343/v1/configuration/object/testObject?UIDARUBA=AOS-COOKIE&config_path=%2Fmm%2Fmynode",
}
`;

exports[`ArubaOsApiClient apiRequest should successfully make a get object request with "getModifiers" 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json;",
    "Cookie": "SESSION=AOS-COOKIE",
  },
  "resolveWithFullResponse": true,
  "strictSSL": true,
  "url": "https://1.2.3.4:4343/v1/configuration/object/testObject?filter=[{\\"acl_sess.accname\\":{\\"$nin\\":[\\"-acl\\"]}},{\\"OBJECT\\":{\\"$eq\\":[\\"acl_sess.accname\\"]}}]&sort=+testKey&type=committed,pending&count=5&offset=5&limit=5&total=5&UIDARUBA=AOS-COOKIE&config_path=%2Fmm%2Fmynode",
}
`;

exports[`ArubaOsApiClient apiRequest should successfully make a set object request if "payload" data is passed 1`] = `
Object {
  "body": "{\\"testProp\\":\\"newValue\\"}",
  "headers": Object {
    "Content-Type": "application/json;",
    "Cookie": "SESSION=AOS-COOKIE",
  },
  "resolveWithFullResponse": true,
  "strictSSL": true,
  "url": "https://1.2.3.4:4343/v1/configuration/object/testObject?UIDARUBA=AOS-COOKIE&config_path=%2Fmm%2Fmynode",
}
`;

exports[`ArubaOsApiClient login should error if the request throws an error 1`] = `[Error: Failed to login. timeout]`;

exports[`ArubaOsApiClient login should successfully login with correct credentials 1`] = `
Object {
  "_global_result": Object {
    "UIDARUBA": "AOS-COOKIE",
    "status": "0",
    "status_str": "You've logged in successfully.",
  },
}
`;

exports[`ArubaOsApiClient logout should error if the request throws an error 1`] = `[Error: Failed to log out timeout]`;

exports[`ArubaOsApiClient logout should successfully logout 1`] = `
Object {
  "_global_result": Object {
    "UIDARUBA": "(null)",
    "status": "0",
    "status_str": "You've been logged out successfully.",
  },
}
`;
